### PR TITLE
Implement XQuery 3.1 arrow operator

### PR DESCRIPTION
**Pull request by wolfgangmm**
Add support for the arrow operator:

``` xquery
('A', 'B', 'C') => count()
```

http://www.w3.org/TR/xquery-31/#id-arrow-operator

### CODE REVIEW

**Comment by joewiz** [comment]
Wonderful!

File: test/src/xquery/xquery3/arrowop.xql [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+declare
+    %test:assertEquals(11)
+function ao:func-in-sequence1() {
+    "Hello" => concat(" world") => string-length()
+};
+
+declare
+    %test:assertEquals("HELLO", "WORLD")
+function ao:func-in-sequence2() {
+    "hello world"=>upper-case()=>normalize-unicode()=>tokenize("\s+")
```
Comment (adamretter) [inline test/src/xquery/xquery3/arrowop.xql]: The test result does not show that `normalize-unicode` was or was not called. Perhaps a better test case would be:
```xquery
declare
    %test:assertEquals("123", "abc", "dl")
function ao:func-in-sequence2() {
    "123 abc ㎗" => upper-case() => normalize-unicode() => tokenize("\s+")
}
````

**Comment by adamretter** [comment]
The following test-case for the arrow operator fails:

```xquery
function($k, $v) {
	"$k=" || $k
} => map:for-each-entry(
	map {
		"a" : "1",
		"b" : "2",
		"c" : "3"
	},
	?
)
```

**Comment by adamretter** [comment]
Also, I think `map:for-each-entry` was renamed to just `map:for-each` https://www.w3.org/Bugs/Public/show_bug.cgi?id=29269. I will send a separate PR for this later...

**Comment by joewiz** [comment]
Just used the arrow operator for the first time by testing this branch. I can already tell I will be using this a lot.

**Comment by joewiz** [comment]
@adamretter Re: your test case above, where do you see in the spec that `?` can be used to identify which parameter the arrow operator passes in?  As I read it http://www.w3.org/TR/xquery-31/#id-arrow-operator states:

> An arrow operator applies a function to the value of a primary expression, using the value **as the first argument to the function**.

[My emphasis.]

**Comment by dizzzz** [comment]
Oops, still not pulled in? @wolfgangmm

**Comment by adamretter** [comment]
@joewiz Yes that is indeed the case. However, I am using `?` to create a partially applied function, so:

```xquery
  map:for-each(
    map {
      "a" : "1",
      "b" : "2",
      "c" : "3"
    },
    ?
  )
```
Results in an anonymous function, that takes a single parameter, and that is why you can then use the arrow `=>` operator to supply its parameter.

**Comment by joewiz** [comment]
@adamretter Given the status of this issue as described at http://markmail.org/message/3whinktocaznfvr4, do you think this PR can be merged so we can begin using this feature for cases where the value of the primary expression is intended to be used as the first argument to the function. Perhaps your use case could be accounted for once the matter is settled by the WG?

**Comment by wolfgangmm** [comment]
@joe the latest version of the spec indeed changes the operator precedence. we'll have to change the grammar to adopt to this.

**Comment by joewiz** [comment]
Ah, okay! For reference, this topic and a test case is described at https://mailman.uni-konstanz.de/pipermail/basex-talk/2016-January/010050.html.

**Comment by adamretter** [comment]
@joewiz I raised 3 issues, only one of them relates to an unsettled matter, the other 2 are gt genuine errors.

File: test/src/xquery/xquery3/arrowop.xql (hunk 1 above)
Comment (joewiz) [inline test/src/xquery/xquery3/arrowop.xql]: Yes, this is definitely better. A small correction: I think "abc" would be upper-cased, so the test case should be

```xquery
declare
    %test:assertEquals("123", "ABC", "dl")
function ao:func-in-sequence2() {
    "123 abc ㎗" => upper-case() => normalize-unicode('NFKC') => tokenize("\s+")
}
```

**Comment by dizzzz** [comment]
a bit quiet here...... for 3.0 or 3.1 ?

**Comment by joewiz** [comment]
I'd love to see this make it into 3.1 if possible? It seems so close!

**Comment by wolfgangmm** [comment]
@joewiz I'm trying to finish it. Test cases pass, but need to align the grammar.

**Comment by adamretter** [comment]
@wolfgangmm cool. I will give it a test after breakfast and all being well I intend to merge it :-)

**Comment by adamretter** [comment]
The first test case (1) I posted:

```xquery
xquery version "3.1";
declare namespace map = "http://www.w3.org/2005/xpath-functions/map";

function($k, $v) {
    "$k=" || $k
} =>
    map:for-each(
        map {
            "a" : "1",
            "b" : "2",
            "c" : "3"
        },
        ?
    )()
```

still fails, although whether it is valid of not is something I should clarify with the XQuery WG. I am happy to report that test cases (2) and (3) that I posted now pass.

I will manually resolve the conflicts and merge...

**Comment by adamretter** [comment]
@wolfgangmm @shabanovd Just a quick note to say, that when you make changes to the XQuery parser, if you can isolate the generated grammar changes to the final commit, then it makes resolving conflicts when people are both working on the parser much much simpler. For example: https://github.com/eXist-db/exist/commit/c9abbef

**Comment by adamretter** [comment]
@wolfgangmm Thank you very much for finishing this, it is a great addition :-)

**Comment by joewiz** [comment]
Agreed! Thank you! This is a great addition to XQuery and eXist!

**Pull request closed without merging**